nat/traversal/pinger.go

```diff
@@ -215,21 +223,33 @@ func (p *Pinger) ping(conn *net.UDPConn) error {
 				n = 128
 			}
 
-			err := ipv4.NewConn(conn).SetTTL(n)
+			err := p.sendPingRequest(conn, n)
 			if err != nil {
-				return errors.Wrap(err, "pinger setting ttl failed")
+				p.eventPublisher.Publish(EventTopic, BuildFailureEvent(stageName, err))
+				return err
 			}
 
-			n++
+			p.eventPublisher.Publish(EventTopic, BuildSuccessEvent(stageName))
 
-			_, err = conn.Write([]byte("continuously pinging to " + conn.RemoteAddr().String()))
-			if err != nil {
-				return err
-			}
+			n++
 		}
 	}
 }
 
+func (p *Pinger) sendPingRequest(conn *net.UDPConn, ttl int) error {
+	err := ipv4.NewConn(conn).SetTTL(ttl)
+	if err != nil {
+		return errors.Wrap(err, "pinger setting ttl failed")
+	}
+
+	_, err = conn.Write([]byte("continuously pinging to " + conn.RemoteAddr().String()))
+	if err != nil {
```

Comment:
It can still be just `return errors.Wrap(err, "pinging request failed")`
if `err` is `nil` `errors.Wrap` will return a nil.